[4.3] Local adapter thumbnails (PR 36552 clone)

obuisard opened the pull request:
### Summary of Changes

This Pull Request is for issue https://github.com/joomla/joomla-cms/issues/36533 [Solves no 2 of the 3 reported problems, the thumbs] and is a rewrite for https://github.com/joomla/joomla-cms/pull/36552 created by @dgrammatiko.

Please refer to the original PR if you want to follow the discussions about the implementation.

The general idea is that image thumbnails are created in the Media Manager and those thumbnails are cached, increasing performance especially when a large amount of images is processed.

The thumbnail creation is 'off' by default.

### Testing Instructions

Go to the _FileSystem - Local_ plugin and set 'Create thumbnails' to 'Yes'  on the 'images' folder.
 
### Actual result BEFORE applying this Pull Request

The media manager loads full size images.

### Expected result AFTER applying this Pull Request

The media manager loads a thumbnail image (max width or height of 200px).

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: https://docs.joomla.org/Chunk4x:Extensions_Plugin_Manager_Edit_FileSystem_Group
- [ ] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/59
- [ ] No documentation changes for manual.joomla.org needed

obuisard committed:
Update plg_filesystem_local.ini

obuisard committed:
Update Image.php

Updated thumbnails creation function

obuisard committed:
Update local.php

Added thumbs to the adapter

obuisard committed:
Update local.xml

Added thumbs field (defaults to true)

obuisard committed:
Update MediaCest.php

Commented out lines in testUpdateFileWithoutAdapter

obuisard committed:
Update LocalAdapter.php

Added thumbs and thumbSize properties
Updated construct, createFile, updateFile, delete, getPathInformation, rglob, checkContent
Added getLocalThumbPaths, getThumb, createThumb functions

obuisard committed:
Update LocalAdapter.php

tab removal

obuisard committed:
Update Image.php

Removed blank space

obuisard committed:
Update MediaCest.php

obuisard committed:
Update LocalAdapter.php

Replaced all use of mkdir() to Folder::create()

obuisard commented:
tests/Codeception/api/com_media/MediaCest.php needs to be updated and fails at this point.

obuisard committed:
Merge branch '4.3-dev' into pr-36552-clone

dgrammatiko commented:
Hey, @obuisard I can't reply on #38722 as Nik has blocked me so I'll write my comment here:
- when I asked you to resurrect this PR I meant to fork it and patch the things that other were asking me to change in the original PR and also fix the failing tests
- - https://github.com/joomla/joomla-cms/pull/36552#pullrequestreview-854742766
- - https://github.com/joomla/joomla-cms/pull/36552#pullrequestreview-854741791
- - https://github.com/joomla/joomla-cms/pull/36552#discussion_r786233867
- About the other PR: the approach on that PR is wrong in the sense that you need a data structure that holds all related info per supported media file (mime type, has preview, allows edit, etc). If the project ends up with different inputs for each of all these *required* data you end up with countless inputs which makes the app really hard to configure and finally use it to its full extend. Mind that since Media manager supports audio and video and other types at some point you might want to provide edit for those (I had a trim plugin for audio and video). Also one more thing: uploads should only allowed if there is a known and matching mime type, the idea that in 2022 people rely on extensions for uploads is a huge security problem. A simple composer require is all it takes: https://packagist.org/packages/league/mime-type-detection to start making things right...

obuisard commented:
Thank you Dimitris @dgrammatiko. I did address the first review, fixed formatting and tested the code successfully. 

As far as the tests are concerned, a solution would be to remove testUpdateFileWithoutAdapter all together, but I would appreciate someone's expertise on this.

dgrammatiko commented:
> a solution would be to remove testUpdateFileWithoutAdapter all together,

@laoneo could you advise here?

laoneo commented:
Removing the test is a bad idea. Better to set the default to 0, so no thumbs will be generated and the admin has to activate it. If it is on by default, then after the update all thumbnails are generated without notice, which can lead to resource issues on large sites.

obuisard committed:
Update local.php

By default, thumbnails are not created

obuisard committed:
Update local.xml

Default the creation of thumbnails to 'no'

obuisard committed:
Merge branch '4.3-dev' into pr-36552-clone

obuisard commented:
I have modified the code to make the thumbnail creation 0 by default. It does solve issues with tests. Thank you @laoneo for the suggestion.

obuisard committed:
Update LocalAdapter.php

Removed the creation of the folder since it is addressed in libraries/src/Image/Image.php as suggested by @Quy in the previous PR

jwaisner commented:
I have tested this item :white_check_mark: successfully on 53bc63a377dfc6b2bb4ddcb040ac284aa49c6e07<br /><br />Everything looks good after applying the PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

N6REJ commented:
RTC

obuisard committed:
Merge branch '4.3-dev' into pr-36552-clone

obuisard commented:
First comment on this PR was edited by the bot, removing information.
Thank you to @dgrammatiko for this work, it's a big step into improving the performance of the media manager.
For information, turning on thumbnails can be found in the Filesystem 'Local' plugin.

![image](https://user-images.githubusercontent.com/5964177/200614905-41144849-e8e2-45d9-b950-3df0a9747f17.png)

obuisard committed:
Merge branch '4.3-dev' into pr-36552-clone

laoneo committed:
Merge branch '4.3-dev' into pr-36552-clone

laoneo commented on libraries/src/Image/Image.php:
```diff
@@ -366,6 +372,24 @@ public function createThumbs($thumbSizes, $creationMethod = self::SCALE_INSIDE,
+     * @throws  \LogicException
+     * @throws  \InvalidArgumentException
+     */
+    public function createThumbs($thumbSizes, $creationMethod = self::SCALE_INSIDE, $thumbsFolder = null)
```
I forgot, here you need to add a deprecated statement and make a pr for the manual.

obuisard replied:
I have added @deprecated, not sure if 5.0, 6.0 or 7.0...

laoneo replied:
Must be 6.0. A pr similar to [this](https://github.com/joomla/Manual/pull/50) one is required as well in the manual.

laoneo commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -267,22 +318,31 @@ public function updateFile(string $name, string $path, $data)
      * @throws  \Exception
      */
-    public function delete(string $path)
+    public function delete(string $path): void
```
```suggestion
    public function delete(string $path)
```
Just to be on the safe side, when somebody extends the LocalAdapter delete function.

laoneo commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -814,4 +873,89 @@ private function getLocalPath(string $path): string
+     * @since   __DEPLOY_VERSION__
+     * @throws  InvalidPathException
+     */
+    private function getLocalThumbPaths(string $path): array
```
Can this be renamed to `getLocalThumbnailPaths` so it is better understandable and more inline with the change in the image class?

laoneo commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -814,4 +873,89 @@ private function getLocalPath(string $path): string
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    private function getThumb(string $path): string
```
Can this be renamed to `getThumbnail` so it is better understandable and more inline with the change in the image class?

laoneo commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -814,4 +873,89 @@ private function getLocalPath(string $path): string
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    private function createThumb(string $path, string $thumbPath): bool
```
Can this be renamed to `createThumbnail` so it is better understandable and more inline with the change in the image class?

laoneo commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -55,22 +55,52 @@ class LocalAdapter implements AdapterInterface
+     *
+     * @since  __DEPLOY_VERSION__
+     */
+    private $thumbs = false;
```
Can this be renamed to `thumbnails` so it is better understandable and more inline with the change in the image class?

laoneo commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -55,22 +55,52 @@ class LocalAdapter implements AdapterInterface
+     *
+     * @since  __DEPLOY_VERSION__
+     */
+    private $thumbSize = [200, 200];
```
Can this be renamed to `thumbnailSize` so it is better understandable and more inline with the change in the image class?

laoneo commented on administrator/language/en-GB/plg_filesystem_local.ini:
```diff
@@ -6,5 +6,6 @@
 PLG_FILESYSTEM_LOCAL="FileSystem - Local"
 PLG_FILESYSTEM_LOCAL_DEFAULT_NAME="Local"
 PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_LABEL="Select directories"
+PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_THUMBS_LABEL="Create thumbnails"
```
Must here no every word start with upper case? And please write here also PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_THUMBNAILS_LABEL.

obuisard replied:
Agree

obuisard committed:
Update plugins/filesystem/local/src/Adapter/LocalAdapter.php

Co-authored-by: Allon Moritz <allon.moritz@digital-peak.com>

obuisard committed:
Update local.xml

Changed label to PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_THUMBNAILS_LABEL

obuisard committed:
Update plg_filesystem_local.ini

Changed PLG_FILESYSTEM_LOCAL_DIRECTORIES_DIRECTORY_THUMBNAILS_LABEL
Uppercased column names

obuisard committed:
Update LocalAdapter.php

Changed thumb to thumbnail where applicable

dgrammatiko commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -293,17 +353,17 @@ public function delete(string $path)
+     * - mime_type:         The mime type
+     * - width:             The width, when available
+     * - height:            The height, when available
+     * - thumbnail_path     The thumbnail path of file, when available
```
You CANNOT do that without breaking B/C: The `thumb_path` is used here: https://github.com/joomla/joomla-cms/blob/9cbe6cb5b204705a2a66b7b774160d78eb783a01/build/media_source/system/js/fields/joomla-media-select.w-c.es6.js#L132

laoneo replied:
Yes you are right, sorry @obuisard , I was wrong here.

obuisard replied:
Yes, I have figured it out while testing it. Thank you both.

richard67 commented:
Back to pending because changes were made and more changes were requested.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

dgrammatiko commented on plugins/filesystem/local/src/Adapter/LocalAdapter.php:
```diff
@@ -346,8 +406,7 @@ private function getPathInformation(string $path): \stdClass
 
-                // @todo : Change this path to an actual thumbnail path
-                $obj->thumb_path = $this->getUrl($obj->path);
+                $obj->thumbnail_path = $this->thumbnails ? $this->getThumbnail($path) : $this->getUrl($obj->path);
```
Same here

laoneo replied:
Again you are right!! Thanks for checking.

obuisard replied:
Yep! Thank you both for the review, I have checked the file again

obuisard committed:
Update LocalAdapter.php

thumb_path should not have been renamed

obuisard committed:
Update Image.php

Added the deprecation on createThumbs

obuisard committed:
Update Image.php

changed deprecated version number

obuisard committed:
Merge branch '4.3-dev' into pr-36552-clone

obuisard commented:
Troy @N6REJ and Jacob @jwaisner, would you mind re-testing this so we can merge it? Thank you so much!

jwaisner commented:
I have tested this item :white_check_mark: successfully on d82abd9fed569dccc9e8772486ae20137d4fd510<br /><br />Everything retested well. No issues.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

N6REJ commented:
I have tested this item :red_circle: unsuccessfully on d82abd9fed569dccc9e8772486ae20137d4fd510<br /><br />in my test the images were already thumbnailed without the PR.  This was " The currently installed Joomla! version is "‎4.3.0-dev"‎4.3.0-dev"
with the latest composer & npm.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

obuisard commented:
> in my test the images were already thumbnailed without the PR. 

That's strange. The thumbnails created with the PR should be 200x200.

dgrammatiko commented:
@obuisard what @N6REJ is showcasing is the SIZE of the container DIV, not the size of the image. To get the size of the image open the browser tools, select the image in the elements tab and in the console type `$0.naturalWidth` (or naturalHeight for the height). False alarm...

obuisard commented:
> @obuisard what @N6REJ is showcasing is the SIZE of the container DIV, not the size of the image. To get the size of the image open the browser tools, select the image in the elements tab and in the console type `$0.naturalWidth` (or naturalHeight for the height). False alarm...

Oh good catch Dimitris @dgrammatiko! @N6REJ can you double-check? Thank you!

Quy commented:
Blurry quality:
![38805-blurry-quality](https://user-images.githubusercontent.com/368084/213527044-ab0984a2-3fb6-4da5-978a-2537f612c6ab.jpg)

Before PR:
![38805-gif-before](https://user-images.githubusercontent.com/368084/213527091-0c91ad2a-46ae-46a3-bda5-7bbd1130b156.jpg)

After PR:
![38805-gif-after](https://user-images.githubusercontent.com/368084/213527124-e425e96e-c0b8-4724-a116-69cfa768eb60.jpg)

obuisard commented:
Thank you, Troy @N6REJ for reporting those issues. 
The image quality and handling of transparency for GIFs is unrelated to this PR, but improvements need to be made to the image library itself. Something for another PR...

N6REJ commented:
I have tested this item :white_check_mark: successfully on d82abd9fed569dccc9e8772486ae20137d4fd510<br /><br />retested as instructed.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38805">issues.joomla.org/tracker/joomla-cms/38805</a>.</sub>

obuisard committed:
Merge branch '4.3-dev' into pr-36552-clone

obuisard commented:
A big thank you to Dimitris @dgrammatiko and the testers !

carlitorweb commented:
**Documentation added** :heavy_check_mark:

dgrammatiko commented:
@obuisard I haven't checked but is the thumbnails functionality bypasses the svg files? If not it needs to do so, the thumbs helper expects only files that the Image class can handle. I'll try to check the next days if I get some free time

obuisard commented:
> @obuisard I haven't checked but is the thumbnails functionality bypasses the svg files? If not it needs to do so, the thumbs helper expects only files that the Image class can handle. I'll try to check the next days if I get some free time

@dgrammatiko I have checked SVG and image thumbnail creation and both work well together

N6REJ commented:
it's sad that we are using GD instead of imagemagick.

obuisard commented:
> it's sad that we are using GD instead of imagemagick.

I know, but GD is more widely installed than imagick. It would be great to offer the use of either one of those graphic libraries in Joomla.

dgrammatiko commented:
You'll be surprised but GD (PHP 8.1+) is (median) faster than imagick. How do I know? I keep testing these for [responsive-images](https://responsive-images.dgrammatiko.dev)

N6REJ commented:
@dgrammatiko faster but I thouight imagick quality was far higher?
